### ISSUE TITLE
Include errors for magnitudes, either native to catalog or as an "add-on" with photo-z's (TBD)

### ISSUE THREAD

**Issue by evevkovacs** [LSSTDESC/cosmodc2#41]
@aphearin @sschmidt23  Sam Schmidt and the PZ group have developed a notebook to compute errors on magnitudes for protoDC2. See [here](https://github.com/LSSTDESC/pz_pdf/pull/4 ). We could adopt this technology to supply magnitude errors for cosmoDC2, or, alternatively, we can work with PZ to make an "add-on" catalog that includes errors, photometric redshifts, etc.  Given that the [code in the notebook](https://github.com/LSSTDESC/pz_pdf/blob/issue/3/add_errs_to_dc2v5/notebooks/proto-dc2_v5_magerrors_no_iterator.ipynb) requires the use of an error model, the latter approach is probably the best.

**Comment by aphearin** [comment]
Magnitude error models in post-processing sounds great! I guess it would only need to be an "add-on" if the computation is a lot slower than the disk I/O; otherwise, probably better to just implement this as a derived quantity function in the GCR framework. Or maybe that's what you meant by "add-on"? 
CC @yymao may have an opinion about this implementation. Details aside, this sounds like a nice feature, so thanks @sschmidt23 and co.!

**Comment by yymao** [comment]
Agreed with @aphearin, if the calculation of magnitude errors is slower than disk I/O, we can make it an "add-on". Otherwise we can just calculate them as derived quantities in the reader.

**Comment by sschmidt23** [comment]
Hi @evevkovacs  and @aphearin, the model is very simple, it's just an implementation of the simple S/N estimates in Ivezic et al (2008) (recently revised this year): https://arxiv.org/abs/0805.2366 See Section 3.2.1 and equations 5, 6, and 7.  Alex Abate wrote code to calculate this in the PhotoZDC1 repo, my notebook basically just updates the numbers from the Science Book values to the 2018 Ivezic numbers and creates a spline interpolation rather than calculating the error for each galaxy, as we weren't planning to vary depth, seeing, etc... Rather, we were just assuming uniform 10 year depth.  
with the spline it only takes about 30 seconds to calculate the ugrizy magnitudes and errors for the ~18 million galaxies in proto-dc2_v5.0.  The estimates are for point source objects, Alex added a fudge factor of just a zero point offset to account for extended objects, I'm sure we could come up with a more realistic scheme for extended objects with a little bit of work.  Again, it's a very simple model, we might want to check around and see if anyone has a better model that could replace this.

Alex's code is here:
https://github.com/LSSTDESC/PhotoZDC1
look at src/photErrorModel.py 

One thing I did definitely want to do is check this error model against the run1.1p coadds to see how they compare.  Because the model does formulate errors in terms of visits, we should be able to do a pretty nice comparison between the model and the errors that the Stack spits out if we have a depth map.  I started looking at how to look at the number of exposures per RA/DEC in run1.1p, but now I see NERSC is down for maintenance for a few days, I'll take another look once things are back up on Tuesday.  If one of you has either a Butler, GCR, or sqlite method for getting number of exposures per band in run1.1p that would be helpful.

**Comment by evevkovacs** [comment]
@sschmidt23 Hi Sam, Thanks a lot. One question: If the model is visit based, and we were thinking of making add-on catalogs for this to the extra-galactic catalogs, does that mean the best we could do is to give some geusstimate of say 1-year errors, 10-year errors etc. In effect, these would be estimates for the co-add errors, right?  As to # of exposures, I haven't played with images so cannot help directly with your last question except suggesting to look in the DC2 tutorials for examples, or asking one of the experts.

**Comment by sschmidt23** [comment]
Yes, the model basically just calculates predicted S/N for LSST given predictions of sky brightness, seeing, airmass, atmospheric effects, instrument noise, etc..., and exposure time.  You can do it on a per-object basis if you know all of those parameters.  This was a big reason I wanted to look at run 1.1p, I'm assuming those quantities are recorded for each of the images and we can see how well the model compares.  I took a quick look at tract 4850 of 1.1p and saw definite RA/DEC dependence on the error vs magnitude in i-band that I assume was due to differing numbers of exposures, once we construct a depth map we can see how well the model compares.  We may also be able to come up with an empirical extension to the model based no simulated images for how to better model extended object errors vs point sources.
